### DISCUSSION TITLE
How do I add toolbar functionality

### DISCUSSION THREAD

**Discussion by zyc329** [imzbf/md-editor-rt#85] [Q&A]
The business requirement has an upload file capability that I want to add but can't seem to find examples where I can throw methods to modify html and add toobars

  业务需求有一个上传文件的功能，我想添加但是好像没有找到可以抛出修改html的方法和添加toobars的例子

**Comment by imzbf** [comment]
有示例：[自定义工具栏](https://imzbf.github.io/md-editor-rt/zh-CN/demo#%F0%9F%92%AA%20%E8%87%AA%E5%AE%9A%E4%B9%89%E5%B7%A5%E5%85%B7%E6%A0%8F)

**Comment by zyc329** [comment]
> 有示例：[自定义工具栏](https://imzbf.github.io/md-editor-rt/zh-CN/demo#%F0%9F%92%AA%20%E8%87%AA%E5%AE%9A%E4%B9%89%E5%B7%A5%E5%85%B7%E6%A0%8F)

 const emojiHandler = (emoji: string) => {
    const generator: InsertContentGenerator = () => {
      return {
        targetValue: emoji,
        select: true,
        deviationStart: 0,
        deviationEnd: 0
      };
    };

    props.onInsert(generator);
  };

封装的时候 主要是targetValue 控制好即可么

**Comment by zyc329** [comment]
> 自定义工具栏
十分感谢您的回复，我希望能再请教你两个问题
1:我希望我点击超链接的时候是 _black 打开新页面下载文件 有提供可以提供修改生成html的实力出来么？
2:我是要做一个上传文件的功能，和您原本的上传图片一样弹出一个框那种，但是我在源码中没有找到如何实现的，我主要是想看看是如何实现onUploadImg是如何注册出来的，有提供配置么

**Comment by imzbf** [comment]
> > 自定义工具栏
> > 十分感谢您的回复，我希望能再请教你两个问题
> > 1:我希望我点击超链接的时候是 _black 打开新页面下载文件 有提供可以提供修改生成html的实力出来么？
> > 2:我是要做一个上传文件的功能，和您原本的上传图片一样弹出一个框那种，但是我在源码中没有找到如何实现的，我主要是想看看是如何实现onUploadImg是如何注册出来的，有提供配置么

第一个参考一下这里：https://github.com/imzbf/md-editor-v3/issues/274
第二个你参考一下`dev-docs`的源码，里面有默认工具栏，下拉菜单工具栏和弹窗工具栏的实际示例。

**Comment by zyc329** [comment]
> 栏

我有看到源码develop分支中component封装的几个组件，但是我没有看到，也看到了您封装的EmojiExtension，但是我没有找到img上传的实现 刚刚去翻了一下dev-dos分支 好像大致内容一样，能告知我一下具体实际事例的路径

**Comment by imzbf** [comment]
> > 栏
> 
> 我有看到源码develop分支中component封装的几个组件，但是我没有看到，也看到了您封装的EmojiExtension，但是我没有找到img上传的实现 刚刚去翻了一下dev-dos分支 好像大致内容一样，能告知我一下具体实际事例的路径

上传的功能需要自己在弹窗里实现，既可以使用UI库的上传组件，也可以自己通过`<input />`标签封装上传文件组件

内部的上传图片也只是做了文件选择这部分，实际的上传逻辑还是开发者自己在处理。

**Comment by zyc329** [comment]
> > > 栏
> > 
> > 
> > 我有看到源码develop分支中component封装的几个组件，但是我没有看到，也看到了您封装的EmojiExtension，但是我没有找到img上传的实现 刚刚去翻了一下dev-dos分支 好像大致内容一样，能告知我一下具体实际事例的路径
> 
> 上传的功能需要自己在弹窗里实现，既可以使用UI库的上传组件，也可以自己通过`<input />`标签封装上传文件组件
> 
> 内部的上传图片也只是做了文件选择这部分，实际的上传逻辑还是开发者自己在处理。

没有办法像这个组件一样组册一个onUploadImg到外面来事件绑定是么

**Comment by imzbf** [comment]
既然都可以在弹窗里添加上传文件组件了，直接监听上传文件的事件，不需要这些额外的东西了
